### PR TITLE
Introduce data mixing recipe yaml files

### PR DESCRIPTION
**Pull request by bbrowning**
This introduces Recipe yaml files, which are used both as an input into the data mixing process and as an output of the process.

As an input, we have some default recipe files that specify any precomputed datasets that should be mixed with data from new skills when generating the overall mix of samples that will be sent to the training process. In our specific case, we're adding in a single precomputed dataset of `instructlab/InstructLabCommunity` that gets pulled from HuggingFace and mixed with new skill samples.

As an output of the data generation process, we write recipe yamls to document which datasets were mixed together and in what proportions along with the system prompt that was used during the generation. Here's an example of a recipe yaml put into the output directory after running data generation:

```yaml
datasets:
- path: instructlab/InstructLabCommunity
  sampling_size: 1.0
- path: node_datasets_2024-07-25T10_27_12/knowledge_tonsils_overview_e2e-tonsils_p10.jsonl
  sampling_size: 1.0
metadata:
  sys_prompt: "I am, Red Hat\xAE Instruct Model based on Granite 7B, an AI language\
    \ model developed by Red Hat and IBM Research, based on the Granite-7b-base language\
    \ model. My primary function is to be a chat assistant."
```

While this change is relatively small, it came from some much larger PRs that were co-authored by multiple people so giving credit to all of those here.

Parts of this are extracted and rebased from
https://github.com/aakankshaduggal/sdg/pull/4
https://github.com/aakankshaduggal/sdg/pull/20

Refs #162, #171, #185, #201.

### CODE REVIEW

**Review by aakankshaduggal — APPROVED** [review]
Thanks! This looks great @bbrowning 🚢

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:428]: Hmm, this is relative to CWD? How do we know the CWD is `instructlab/sdg` ? What when `ilab` is consuming the library as a wheel?

For loading pipeline configs, we load it from the Python package: https://github.com/instructlab/sdg/blob/cbee53c2453876f2602c86ef3e5a803689604716/src/instructlab/sdg/generate_data.py#L229

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:104]: Make recipe_path optional, default to None?

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:106]: Handle recipe_path = None

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:116]: I guess I'd imagine this being the reverse? The one in the generation pipeline would be the default if there's none specified in the file?

File: src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml
Comment (markmc) [inline src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml:1]: Let's just remove this empty file and just create an empty Recipe() in the code?

**Review by markmc — CHANGES_REQUESTED** [review]
Some minor comments, but I think the point about what `recipe_path` is relative to is probably important - e.g. has this been tested with `ilab data generate` ?

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:43]: You know, it's probably important we support relative paths here too?

e.g. if our default recipe referenced a `.jsonl` file as a dataset, we'd need it to be relative to some install path

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:428]: That's a great catch on the CWD, as I know I've only tested from within an instructlab/sdg clone folder and I suspect that's how others are testing, which is why this happens to work as-is but would not reliably work. I do think we should be able to load and reference these files like the configs instead of the more naive way here.

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:43]: We might be able to assume these are always absolute paths at the moment, as the recipes are written as absolute paths and any downstream usage of this I've heard assumed that these files would be in a known absolute path. But you bring up a good point that we should test that assumption, as well as an overall point about portability of the recipe file if the paths to jsonl files aren't relative to the data generation output directory, at least for the generated recipe yaml files and outputs.

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:116]: Ok, after some back-and-forth discussion, it sounds like we don't really ever need to load the system prompt from these recipe files. It gets written to them to document what was used, but we don't need to load it back in. I'll make that change and unconditionally replace this with the system prompt passed in to the function call.

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:43]: I had some offline discussion with the team and we agreed it likely makes more sense for these to be relative paths, relative to the generation output directory. So, my example recipe output above would change from:

```yaml
datasets:
- path: instructlab/InstructLabCommunity
  sampling_size: 1.0
- path: /home/my_user/.local/share/instructlab/datasets/node_datasets_2024-07-24T15_07_58/knowledge_tonsils_overview_e2e-tonsils_p10.jsonl
  sampling_size: 1.0
sys_prompt: "I am, Red Hat\xAE Instruct Model based on Granite 7B, an AI language\
  \ model developed by Red Hat and IBM Research, based on the Granite-7b-base language\
  \ model. My primary function is to be a chat assistant."
```
to:

```yaml
datasets:
- path: instructlab/InstructLabCommunity
  sampling_size: 1.0
- path: node_datasets_2024-07-24T15_07_58/knowledge_tonsils_overview_e2e-tonsils_p10.jsonl
  sampling_size: 1.0
sys_prompt: "I am, Red Hat\xAE Instruct Model based on Granite 7B, an AI language\
  \ model developed by Red Hat and IBM Research, based on the Granite-7b-base language\
  \ model. My primary function is to be a chat assistant."
```

For robustness sake, they should probably accept either absolute or relative with the relative option becoming relative to the output directory.

**Comment by bbrowning** [comment]
Marking as draft while I work on absolute vs relative paths a bit, as well as address the other feedback here.

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [reply src/instructlab/sdg/datamixing.py:103]: That's pretty bizarre too though ... you load a config file and some of it gets ignored?

Maybe add a 'metadata' section (or similar) to hold `sys_prompt` ... the semantics would be that anything in that section is info written out by the mixer, doesn't influence the mixing process?

File: src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml
Comment (markmc) [inline src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: This worries me a bit ... downloading stuff from Hugging Face, down in the depths of the library

UX-wise, it feels a little like 'ilab model download' ... something you would do as part of setting up your workflow

It's like, you could plausibly do `data generate` and `model train` iterations on your laptop on a plane with no Internet, except for this?

Consider also a unit test like #208 ... all of this datamixing code is self-contained, apart from this external dependency on Hugging Face?

If we had some sort if `ilab data setup` command ... and it downloaded this dataset into local cache ... what else might that command do? e.g. validate that the teacher models required are available from the configured OpenAI endpoint?

I'm not really suggesting this as a blocker for merging ... but I would like to file an issue which acknowledges that this isn't ideal and has some sort of plan for resolving it

Comment (bbrowning) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: I agree and share some of these concerns, as it's a download that just happens behind the scenes automatically. I know some downstream users intend to replace this HuggingFace with a file on disk as the preconfigured dataset, which is why the code supports loading from either. It's a 32MB file, which is not necessarily too large to just include as part of our python package, although that may not be ideal either.

I'd prefer to not introduce a new command for this right now, as depending on how we address this it may not require a separate command to download (like if we just ship it in some python package). Or, it may need a separate command. Either way I think a proper fix warrants a longer discussion than what we'll want for this release.

Is there a better alternative we should go with for this release? Do you think just placing the file directly in the git repository and including it as part of the SDG package is a less bad option?

Comment (markmc) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: A 32Mb file is not ideal, but in the context of some of these other packages ...

Paging @tiran @dhellmann @n1hility

File: src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml
Comment (markmc) [reply src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml:1]: Downstream users would need to change the contents of the package to change this recipe file anyway, so I don't think it's a problem to also change 1 line of code

If we really want to support downstream recipes, this is not how to do it - and we should file an issue for that requirement

I really don't want to leave stuff like this in place - someone upstream is quite likely to remove it, not understanding the downstream consequences. And if downstream no longer needs it, they're very unlikely to notify upstream of this.

That's how we end up with cruft that people are afraid to touch. Not a huge deal in this case, but broken windows and all that

Comment (dhellmann) [reply src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml:1]: Downstream we would be unlikely to put any default data file in the code package. Instead, we would install it separately as part of building the application image.

File: src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml
Comment (tiran) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: > download that just happens behind the scenes automatically

To a security engineer, this is a horror story in eight words. CC:  @sidhpurwala-huzaifa 

How is the download triggered? How is the content tracked, verified and signed? Please assume that $user wants to run InstructLab on an air-gapped system, with all data pre-loaded and without any access to the internet. Some users want to use InstructLab with PII or confidential data.

Very recently we had a similar behavior in a 3rd party dependency. It triggered a HTTPS request and a download every time we ran `ilab` command. The code ran on `import` time.

I would prefer a Python package or another mechanism to deliver and track the file.

Comment (dhellmann) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: Generally, libraries should not introduce side-effects like downloading files and writing to potentially protected locations on disk.

More specifically, when we get to the point of running tests for individual packages before using them downstream, those tests will have to run without accessing the internet. Everything they need will have to be installed before the tests run.

Comment (bbrowning) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: This entire discussion makes sense. Are there suggestions for how we do want to ship this precomputed dataset in upstream? How we want to reference it? And, how would downstream users supply their own precomputed dataset (replacing/obviating the upstream one specified here) in their own releases?

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:105]: That's not a bad idea. I'm glad these discussions are happening, although it's starting to become clear this should have happened a couple of weeks ago when these features were first being developed elsewhere.

Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:105]: The `sys_prompt` only gets used when writing out the mixed dataset, where we overwrite the system prompt of every source sample to be the prompt specified here. And the reason we include it in the Recipe yaml file is so there's a record of what system prompt exists in the mixed dataset without having to go look inside the dataset file itself. So, maybe when loading a Recipe from disk or when writing it out we should always include the sys_prompt, and take the sys_prompt out of the constructor here. So the constructor would have less logic - loads a recipe from disk or doesn't, but none of this logic about which sys_prompt to use.

Then, within our own code, we'd change the sys_prompt of the recipe after loading it from disk, so that it has the system prompt we're using during generation.

Instead of:
```python
        self.knowledge_recipe = Recipe(
            recipe_path="src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml",
            sys_prompt=self.sys_prompt,
        )
```

we'd move to something like:
```python
        self.knowledge_recipe = Recipe(
            recipe_path="src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml",
        )
        self.knowledge_recipe.sys_prompt = self.sys_prompt
```

This would perhaps be cleaner and make more sense? It removes the overwriting of sys_prompt concerns out of the Recipe class and lets that just focus on loading/dumping itself. And it still lets us override the sys_prompt in our usage where we want to force the system prompt to be what we ran the generation pipeline with, regardless of what was or was not specified in some default recipe.

Comment (markmc) [reply src/instructlab/sdg/datamixing.py:105]: Isn't that the same as just not reading `sys_prompt` from the recipe file at all - e.g. by writing it out in a `metadata` section, and just not reading that section when loading?

**Comment by markmc** [comment]
Ok, lots of great discussion in these threads. It's a lot to absorb, but I think it boils down to a pretty straightforward set of changes

**Default recipes - downstream**

If a downstream user/packager wants to add default recipes (and datasets), they should install them in `/usr/share/instructlab/sdg`

- `generate_data()` should supply `default_recipes_dir` path to `DataMixer` using `platformdirs.PlatformDirs`
- downstream users should install recipes into `/usr/share/instructlab/sdg/default_data_recipes/{knowledge,skills}.yaml`
- downstream users should install datasets into `/usr/share/instructlab/sdg/datasets`
- this a packaging thing - e.g. it would typically be done as part of the container image build
- the library would use any default recipes found there

**Default recipes - testing**

Since we won't have default recipes in the library, we do need some way of ensuring this code path is tested - this can be done with a unit test. But it is a requirement to merge this PR IMO.

**Community dataset recipe - upstream**

Assuming we have at least unit testing of recipes, I don't think this is a blocker to merging this PR - I'm fine with just filing an issue to capture this.

We want to enable upstream users to mix the `instructlab/InstructLabCommunity` dataset.

They would download the dataset and place it in `~/.cache` ? Potentially there would be an `ilab data download` command to do this

It can be supplied with e.g. `ilab data generate --skills-dataset=` and ilab would construct a simple skills recipe (in memory) and pass it to the library

This path could be tested in our e2e CI, which would also improve automated testing coverage of the library.

**Hugging Face downloads in library**

Given the above, there is no need for the library to have support for downloading from Hugging Face

**Relative paths**

There is no requirement to support relative paths to recipes files.

Paths to dataset should be considered relative to the recipe file.

**Empty recipes**

`Recipe()` should be sufficient to create an empty recipe.

**sys_prompt**

It sounds like we never want to use the `sys_prompt` value from a user-supplied recipe. It is included in recipe files that are written by the library for information only.

Move `sys_prompt` in the recipe format to a new `metadata` section, and ignore that whole section when reading.

**Comment by bbrowning** [comment]
Thank you @markmc , that is super helpful. I've pushed a commit that should adjust to the `sys_prompt` guidance under a new `metadata` section, but it will take a bit more time/thought to implement some of the other changes. If you feel comfortable with the downstream dataset chunk of this or anything else, feel free to grab it. Otherwise I'll dig in to them in a bit.

**Comment by bbrowning** [comment]
For reference, the generated recipe yaml files now look like this:
```yaml
datasets:
- path: instructlab/InstructLabCommunity
  sampling_size: 1.0
- path: node_datasets_2024-07-25T10_27_12/knowledge_tonsils_overview_e2e-tonsils_p10.jsonl
  sampling_size: 1.0
metadata:
  sys_prompt: "I am, Red Hat\xAE Instruct Model based on Granite 7B, an AI language\
    \ model developed by Red Hat and IBM Research, based on the Granite-7b-base language\
    \ model. My primary function is to be a chat assistant."
```

I've updated the original comment at the top of this PR to reflect the current state as well.

File: src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml
Comment (dhellmann) [reply src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml:1]: I'd really rather not assume we're going to manage patches downstream for packages where we have enough influence to ensure the design doesn't require it. We have enough work doing that for packages where we have no influence over the maintainers.

File: src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml
Comment (dhellmann) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml:2]: These aren't the only inputs the system needs, right? Is there a pattern to follow in how things like models are made available for users?

Comment (markmc) [reply src/instructlab/sdg/configs/skills/data_recipe/default_recipe.yaml]: See https://github.com/instructlab/sdg/pull/203#issuecomment-2250444499

File: src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml
Comment (markmc) [reply src/instructlab/sdg/configs/knowledge/data_recipe/default_recipe.yaml]: See https://github.com/instructlab/sdg/pull/203#issuecomment-2250444499

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py]: Resolving this, as the code has evolved significantly from here and addresses this concern.

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py]: Resolved by now writing the sys_prompt out to a metadata section.

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py]: Loading datasets from paths relative to the recipe file is now supported.

**Comment by bbrowning** [comment]
Assuming CI passes, I believe all concerns are addressed now.

@markmc  added in support for loading of default recipe files from `platformdirs.PlatformDirs`, allowing downstream users to package default recipe yaml files in a place like `/usr/share/instructlab/sdg/default_data_recipes/{knowledge,skills}.yaml`. I added a unit test to ensure that if we find default recipe yaml files in those directories, they get loaded and used.

#201 was opened and expanded with more details as a result of the conversations in this PR to figure out how to properly get the precomputed dataset from HuggingFace into the upstream `ilab` experience in an easy-to-use way. This may involve a new command, such as `ilab data download`, but discussion around that design can happen in that issue.

This PR no longer includes downloading anything from HuggingFace, as several valid concerns were brought up about doing that silently and without explicit user intent.

Empty `Recipe()`s can be created, and we have unit tests ensuring that works along with a few other permutations.

We do not load sys_prompt from the recipe files, but do write it out to them under a new `metadata` section so that it's still obvious which system prompt was used during data mixing without having to look inside the data files themselves.

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:125]: I thought we were removing this?

Comment (markmc) [reply src/instructlab/sdg/datamixing.py:125]: Hugging Face download mentioned in docstring too

Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:125]: Yes, we are! However in my multi-tasking it looks like I accidentally did that in a different local branch than the one I pushed here. Need more coffee, will fix as I have a rebase conflict to fix anyway.

File: src/instructlab/sdg/datamixing.py
Comment (markmc) [inline src/instructlab/sdg/datamixing.py:526]: Maybe I'm being dumb, but I really don't understand what's going on here

We've just written a recipe file to a path, and setting `recipe.recipe_path` is going to influence where the dataset gets written? But `save_mixed_dataset()` doesn't use `recipe_path`?

Thanks!

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py]: Sorry, fixed that now.

**Review by markmc — APPROVED** [review]
Looks good to me, thanks @bbrowning 

Happy for you to merge this when CI passes

File: src/instructlab/sdg/datamixing.py
Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py:526]: `save_mixed_dataset()` DOES use `recipe_path`, as it reads the datasets in from disk to mix them. I wasn't very happy with this when I wrote it, and I should have recognized the code smell myself when I wrote that comment explaining what it's doing. What happens is we loaded a default recipe from whatever path - say `/default_recipes/skills.yaml` (or instantiated an empty `Recipe()`, perhaps). Then, we save the recipe to a new path - like `/generated_data/skills_train_recipe.yaml`. At this point, the actual Recipe instance we have still thinks its `recipe_path` is `/default_recipes/skills.yaml`, even though we wrote it out to a different path. The fix staring me in the face as I type this is that when someone calls `save_recipe`, that should update the Recipe's `recipe_path` to the new path it just wrote itself out to.

The datasets are added to the recipe with a relative path, the recipe is written to a new path, and we then need to update the recipe to know its new path so that it can successfully load the datasets from the relative path, which is relative to the newly written location versus the original location.

Comment (bbrowning) [reply src/instructlab/sdg/datamixing.py]: I pushed one more commit that adjusts this that feels better for me, and keeps this detail encapsulated within the Recipe class instead of leaking out to consumers.

**Comment by bbrowning** [comment]
Thanks! Will squash commits, unmark from Draft, and merge when CI is good.

**Pull request merged**
